plugin/evm/header/dynamic_fee_windower.go

```diff
@@ -0,0 +1,261 @@
+	if err != nil {
+		return ap3.Window{}, err
+	}
+
+	if timestamp < parent.Time {
+		return ap3.Window{}, fmt.Errorf("cannot calculate fee window for timestamp %d prior to parent timestamp %d",
+			timestamp,
+			parent.Time,
+		)
+	}
+	timeElapsed := timestamp - parent.Time
+
+	// Add in parent's consumed gas
+	var blockGasCost, parentExtraStateGasUsed uint64
+	switch {
+	case config.IsApricotPhase5(parent.Time):
+		// blockGasCost is not included in the fee window after AP5, so it is
+		// left as 0.
+
+		// At the start of a new network, the parent may not have a populated
+		// ExtDataGasUsed.
+		if parent.ExtDataGasUsed != nil {
+			parentExtraStateGasUsed = parent.ExtDataGasUsed.Uint64()
+		}
+	case config.IsApricotPhase4(parent.Time):
+		// The blockGasCost is paid by the effective tips in the block using
+		// the block's value of baseFee.
+		//
+		// Although the child block may be in AP5 here, the blockGasCost is
+		// still calculated using the AP4 step. This is different than the
+		// actual BlockGasCost calculation used for the child block. This
+		// behavior is kept to preserve the original behavior of this function.
+		blockGasCost = BlockGasCostWithStep(
+			parent.BlockGasCost,
+			ap4.BlockGasCostStep,
+			timeElapsed,
+		)
+
+		// On the boundary of AP3 and AP4 or at the start of a new network, the
+		// parent may not have a populated ExtDataGasUsed.
+		if parent.ExtDataGasUsed != nil {
+			parentExtraStateGasUsed = parent.ExtDataGasUsed.Uint64()
+		}
+	default:
+		blockGasCost = ap3.IntrinsicBlockGas
+	}
+
+	// Compute the new state of the gas rolling window.
+	dynamicFeeWindow.Add(parent.GasUsed, parentExtraStateGasUsed, blockGasCost)
+
+	// roll the window over by the timeElapsed to generate the new rollup
+	// window.
+	dynamicFeeWindow.Shift(timeElapsed)
+	return dynamicFeeWindow, nil
+}
+
+// selectBigWithinBounds returns [value] if it is within the bounds:
+// lowerBound <= value <= upperBound or the bound at either end if [value]
+// is outside of the defined boundaries.
+func selectBigWithinBounds(lowerBound, value, upperBound *big.Int) *big.Int {
+	switch {
+	case lowerBound != nil && value.Cmp(lowerBound) < 0:
+		return new(big.Int).Set(lowerBound)
+	case upperBound != nil && value.Cmp(upperBound) > 0:
+		return new(big.Int).Set(upperBound)
+	default:
+		return value
+	}
+}
+
+func parseFeeWindow(bytes []byte) (ap3.Window, error) {
+	if len(bytes) < FeeWindowSize {
+		return ap3.Window{}, fmt.Errorf("%w: expected at least %d bytes but got %d bytes",
+			errDynamicFeeWindowInsufficientLength,
+			FeeWindowSize,
+			len(bytes),
+		)
+	}
+
+	var window ap3.Window
+	for i := range window {
+		offset := i * wrappers.LongLen
+		window[i] = binary.BigEndian.Uint64(bytes[offset:])
+	}
+	return window, nil
+}
+
+func feeWindowBytes(w ap3.Window) []byte {
+	bytes := make([]byte, FeeWindowSize)
+	for i, v := range w {
+		offset := i * wrappers.LongLen
+		binary.BigEndian.PutUint64(bytes[offset:], v)
+	}
+	return bytes
+}
```

Comment:
The binary representation of the window is in this package rather than the `ap3` package because the binary representation of the acp176 format will essentially need to be in this package to make any sense... And it felt weird for them not to be co-located.